### PR TITLE
Carson Jardine IC 4

### PR DESCRIPTION
**Pull request by carson-jardine**
def find_all_interested_patrons(exhibit)
    @patrons.find_all do |patron|
      patron.interests.include?(exhibit.name)
    end
  end

  def patrons_by_exhibit_interest
    patrons_by_interests = {}
    @exhibits.each do |exhibit|
      patrons_by_interests[exhibit] = find_all_interested_patrons(exhibit)
    end
    patrons_by_interests
  end

  def ticket_lottery_contestants(exhibit)
      patrons_by_exhibit_interest[exhibit].find_all do |patron|
        patron.spending_money < exhibit.cost
      end
  end

This is how I would have refactored museum.rb if I had the time, my methods are too ugly for me not to post this now lol.

### CODE REVIEW

**Comment by KathleenScriver** [comment]
Thanks for the refactor! :) Great job with the commits.

File: lib/museum.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    @exhibits = []
+    @patrons = []
+  end
+
+  def add_exhibit(exhibit)
+    @exhibits << exhibit
+  end
+
+  def recommend_exhibits(patron)
+    @exhibits.find_all do |exhibit|
```
Comment (KathleenScriver) [inline lib/museum.rb:14]: Great job using the find_all enumerable here!

File: lib/museum.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+  def admit(patron)
+    @patrons << patron
+  end
+
+  def patrons_by_exhibit_interest
+    patrons_by_interests = {}
+    @exhibits.flat_map do |exhibit|
+      patrons_by_interests[exhibit] = []
+      @patrons.find_all do |patron|
```
Comment (KathleenScriver) [inline lib/museum.rb:27]: If all you're using your enumerables for is to access data and then put it into a new structure (the new hash), then you only need to use a .each.

File: lib/museum.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        if patron.interests.include?(exhibit.name)
+          if patrons_by_interests[exhibit]
+            patrons_by_interests[exhibit] << patron
+          else
+            patrons_by_interests[exhibit] = [patron]
+          end
+        end
+      end
+    end
+    patrons_by_interests
```
Comment (KathleenScriver) [inline lib/museum.rb:37]: Great, this works!

File: lib/museum.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  end
+
+  def ticket_lottery_contestants(exhibit)
+    patrons_by_exhibit_interest.flat_map do |interest, patrons|
+      if interest == exhibit
+        patrons.find_all do |patron|
+          patron.interests.include?(exhibit.name) && patron.spending_money < exhibit.cost
+        end
+      end
+    end.compact
```
Comment (KathleenScriver) [inline lib/museum.rb:47]: This works but is slightly overly complex :)

File: lib/museum.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  end
+
+  def draw_lottery_winner(exhibit)
+    winner = ticket_lottery_contestants(exhibit).sample
+    unless winner == nil
```
Comment (KathleenScriver) [inline lib/museum.rb:52]: Great job checking for the nil value.

File: test/museum_test.rb [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    @patron_1.add_interest("Gems and Minerals")
+    @patron_2.add_interest("Dead Sea Scrolls")
+    @patron_3.add_interest("Dead Sea Scrolls")
+
+    @dmns.admit(@patron_1)
+    @dmns.admit(@patron_2)
+    @dmns.admit(@patron_3)
+
+
+    refute_equal "Sally", @dmns.draw_lottery_winner(@dead_sea_scrolls)
```
Comment (KathleenScriver) [inline test/museum_test.rb:119]: Hmm, just because it doesn't equal Sally doesn't mean it is equal to Bob or Johnny.

File: test/museum_test.rb [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    assert_nil @dmns.draw_lottery_winner(@gems_and_minerals)
```
Comment (KathleenScriver) [inline test/museum_test.rb:120]: Great job checking for the nil value.

File: test/museum_test.rb (hunk 6 above)
Comment (carson-jardine) [reply test/museum_test.rb:119]: struggled a lot with this one haha I tried to do assert_includes in juuuust about every way except for assert_inlcudes ["Johnny", "Bob"]. I'll know for next time :)

**Pull request closed without merging**